dao/src/main/java/org/thingsboard/server/dao/timeseries/CassandraBaseTimeseriesDao.java

```diff
@@ -159,19 +174,36 @@ public ListenableFuture<Integer> save(TenantId tenantId, EntityId entityId, TsKv
         ttl = computeTtl(ttl);
         int dataPointDays = tsKvEntry.getDataPoints() * Math.max(1, (int) (ttl / SECONDS_IN_DAY));
         long partition = toPartitionTs(tsKvEntry.getTs());
+        String entityType = entityId.getEntityType().name();
+        UUID entityIdId = entityId.getId();
+        String entryKey = tsKvEntry.getKey();
+        long ts = tsKvEntry.getTs();
         DataType type = tsKvEntry.getDataType();
+        BoundStatementBuilder stmtBuilder;
         if (setNullValuesEnabled) {
-            processSetNullValues(tenantId, entityId, tsKvEntry, ttl, futures, partition, type);
-        }
-        BoundStatementBuilder stmtBuilder = new BoundStatementBuilder((ttl == 0 ? getSaveStmt(type) : getSaveTtlStmt(type)).bind());
-        stmtBuilder.setString(0, entityId.getEntityType().name())
-                .setUuid(1, entityId.getId())
-                .setString(2, tsKvEntry.getKey())
-                .setLong(3, partition)
-                .setLong(4, tsKvEntry.getTs());
-        addValue(tsKvEntry, stmtBuilder, 5);
-        if (ttl > 0) {
-            stmtBuilder.setInt(6, (int) ttl);
+            Boolean booleanValue = tsKvEntry.getBooleanValue().orElse(null);
+            String strValue = tsKvEntry.getStrValue().orElse(null);
+            Long longValue = tsKvEntry.getLongValue().orElse(null);
+            Double doubleValue = tsKvEntry.getDoubleValue().orElse(null);
+            String jsonValue = tsKvEntry.getJsonValue().orElse(null);
+            if (ttl == 0) {
+                stmtBuilder = new BoundStatementBuilder(prepare(INSERT_QUERY_WITH_NULL)
```

Comment:
I would suggest you to follow the pattern as before.
I see the `prepare()` here. Before it was a dedicated method that used locks and stored the statement into a volatile variable.
The difference is getting from a variable is much cheaper than searching in a hashmap with computeIfAbsent.
This is a very hot piece of code.
Please, see examples `getSaveStmt(type) getSaveTtlStmt()`